### DISCUSSION TITLE
RAM usage of Kafka

### DISCUSSION THREAD

**Discussion by saloni-2404** [strimzi/strimzi-kafka-operator#4194] [General]
Hello Jakub, happy new year.
Here is a question, why Kafka could be using too much RAM, and what will happen if the RAM limit is breached?

**Comment by scholzj** [comment]
TBH, I'm not sure I have an answer. I do not really understand how these things work. But for a start you would probably need to clarify what do you mean with RAM and limits. Do you mean the Kubernetes requests and limits? Or do you mean more things like the Java heap etc.?

**Reply by saloni-2404** [reply]
What I observed is when the no. of messages into Kafka are getting increased the RAM for Kafka (Kubernetes) in use is also getting increased 
and it breached the requested RAM and it should not be happening as Kafka stores records into PV not in RAM ?

**Reply by scholzj** [reply]
As I said - I'm definitely not an expert on this - but there are different kinds of memory used by Kafka. 

First, you have the Java memory. That consists partially of the Java Heap, but there are also other types of memory it will be using. To large extent, you can control how much heap memory is used with the -Xmx option. But always keep in mind that there will be significant overhead over what you specify as the -Xmx number.

The other big part of memory is the disk / page cache. You are right that Kafka stores records on the PVs. But the things written to the PV or read from it are kept in the disk cache. That is important because it allows Kafka to provide lot of the data from the cache instead of from the disks which is much faster. And that makes it performant.

So in general yes, busy cluster would probably take as much memory as it can take. The cache in general would use any free memory and free it only when someone else requires it.

So, you would need to figure out what kind of memory you are looking at, how did you configured the memory request / limits for the different pods, which part of it breached the limits etc.

**Reply by saloni-2404** [reply]
That helps a lot, Jakub. Thank you.

**Reply by saloni-2404** [reply]
Jakub,
Scenario 1: After 5 days of monitoring we saw that Kafka is using more than requested RAM, so we increased RAM to 18GB and at that time it was using 16GB RAM.
New Scenario: After 6 days of monitoring we are seeing now Kafka is using more than 18GB RAM.

NOTE: There is no increase in load, it is the same from past 6 days, messages are getting pushed and getting consumed from Kafka. Currently there are no messages in Kafka topic, please see attached image and still it is using more than requested RAM.

Can you please suggest what size RAM should be configured for Kafka?

**Reply by scholzj** [reply]
I think you still need to clarify what kind of RAM do you mean and how are you measuring it. The RAM consumed by the disk cache will be getting bigger and bigger as long as there is free memory. So if this is included, then this is no real surprise. Also, when you talk about requested RAM ... does that mean the `request.memory`? Or what do you mean with it? Assuming this is `request.memory`, what is your configured limit? The request is something what can be easily breached. So it is not unexpected if it consumes more.

**Reply by Gulnaz2021** [reply]
Hi Jakub, 

For the above scenario, memory has been defined in the kafka.yaml file in the resource section as :
      resources:
        requests:
          memory: 18Gi
          cpu: "4"
        limits:
          memory: 20Gi
          cpu: "8"

And the above mentioned utilization pattern has been observed.

**Reply by Gulnaz2021** [reply]
And the memory utilization is being tracked from GCP console -> Kubernetes engine -> Workloads

**Reply by scholzj** [reply]
I'm afraid I do not know what exactly GCP console uses for the number - so that does not help me. But most probably sum of everything?

With request 18Gi and limit 20Gi, it is not unexpected that the process will consume more than 18Gi of memory. It would be unexpected if it consumed more than 20Gi. But that might again depend on how exactly do you calculate the number.

**Reply by saloni-2404** [reply]
So Jakub, ideally for 200 topic partitions how much RAM to be specified is recommended?

**Reply by scholzj** [reply]
I don't think that can be answered like this. That depends on many factors.

**Reply by Gulnaz2021** [reply]
Hello Jacob,
Can you please help with the information that, the resource limit the we define in kafka.yaml, suppose for memory. Then is it that the Strimzi operator would inform Kafka internally the limit of the available resource and that it should not exceed the limit, example for memory. So, that Kafka would try to consume all the available memory, but would not cross the limit.

**Reply by scholzj** [reply]
Not really. Java has its own container awareness and is able to detect this on its own from the container it self. So there is nothing like passing to Kafka some option `--memory=8Gi` or something like that. We set it in the Pod / container and Java detects it from there. There are some additional options like the `-Xmx` etc., which tell the Java VM about how the memory should be used as well.

**Reply by Gulnaz2021** [reply]
So, suppose for the below example, that we would configure in kafka.yaml file:
resources:
requests:
memory: 18Gi
cpu: "4"
limits:
memory: 20Gi
cpu: "8",
Then how would Kafka know, that memory limit is 20 Gi. 
And what we've observed, is that for the Kafka cluster, memory utilization has always been near about the memory limit, but has never been crossed. Even with the reduced memory, similar pattern has been observed. So, is it that expected behavior of Kafka to try to use as much of the available memory.

**Reply by scholzj** [reply]
The available memory can be queried from inside the container. It is just a system information as any other. So when Java starts with your configuration, it sees that it has 20Gi of available memory.
